storage/src/nodestore/hash.rs

```diff
@@ -170,49 +155,36 @@ where
                 path_prefix.0.truncate(original_length);
             }
         }
+
         // At this point, we either have a leaf or a branch with all children hashed.
         // if the encoded child hash <32 bytes then we use that RLP
-
         #[cfg(feature = "ethhash")]
-        // if we have a child that is the only child of an account branch, we will hash this child as if it
-        // is a root node. This means we have to take the nibble from the parent and prefix it to the partial path
-        let hash = if let Some(nibble) = fake_root_extra_nibble {
-            let mut fake_root = node.clone();
-            trace!("old node: {fake_root:?}");
-            fake_root.update_partial_path(Path::from_nibbles_iterator(
-                std::iter::once(nibble).chain(fake_root.partial_path().0.iter().copied()),
-            ));
-            trace!("new node: {fake_root:?}");
-            hash_node(&fake_root, path_prefix)
-        } else {
-            hash_node(&node, path_prefix)
-        };
-
+        let hash = Self::compute_node_ethhash(&node, path_prefix, account_child_no_peers);
         #[cfg(not(feature = "ethhash"))]
         let hash = hash_node(&node, path_prefix);
 
         Ok((SharedNode::new(node).into(), hash))
     }
 
     #[cfg(feature = "ethhash")]
+    /// This function computes the ethhash of a single node assuming all its children are hashed.
+    /// The function appends to `path_prefix` and then truncate it back to the original length - we only reuse the memory space to avoid allocations
     pub(crate) fn compute_node_ethhash(
         node: &Node,
-        path_prefix: &Path,
-        have_peers: bool,
+        path_prefix: &mut Path,
+        account_child_no_peers: bool,
     ) -> HashType {
-        if path_prefix.0.len() == 65 && !have_peers {
+        if path_prefix.0.len() == 65 && account_child_no_peers {
             // This is the special case when this node is the only child of an account
             //  - 64 nibbles for account + 1 nibble for its position in account branch node
             let mut fake_root = node.clone();
+            let extra_nibble = path_prefix.0.pop().expect("path_prefix not empty");
```

Comment:
It seems that we do not need to explicitly remove the extra nibble and `hash_node` will only take the first 32 bytes. @rkuris can you confirm?